connect.go

```diff
@@ -17,80 +17,96 @@
 package edgedb
 
 import (
-	"context"
 	"fmt"
+	"sync"
 
-	"github.com/edgedb/edgedb-go/protocol/buff"
-	"github.com/edgedb/edgedb-go/protocol/message"
+	"github.com/edgedb/edgedb-go/internal/buff"
+	"github.com/edgedb/edgedb-go/internal/message"
 	"github.com/xdg/scram"
 )
 
-func (c *baseConn) connect(ctx context.Context, cfg *connConfig) error {
-	buf := buff.New(nil)
-	buf.BeginMessage(message.ClientHandshake)
-	buf.PushUint16(0) // major version
-	buf.PushUint16(8) // minor version
-	buf.PushUint16(2) // number of parameters
-	buf.PushString("database")
-	buf.PushString(cfg.database)
-	buf.PushString("user")
-	buf.PushString(cfg.user)
-	buf.PushUint16(0) // no extensions
-	buf.EndMessage()
-
-	if err := c.writeAndRead(ctx, buf.Unwrap()); err != nil {
+func (c *baseConn) connect(r *buff.Reader, cfg *connConfig) error {
+	c.writer.BeginMessage(message.ClientHandshake)
+	c.writer.PushUint16(0) // major version
+	c.writer.PushUint16(8) // minor version
+	c.writer.PushUint16(2) // number of parameters
+	c.writer.PushString("database")
+	c.writer.PushString(cfg.database)
+	c.writer.PushString("user")
+	c.writer.PushString(cfg.user)
+	c.writer.PushUint16(0) // no extensions
+	c.writer.EndMessage()
+
+	if err := c.writer.Send(c.conn); err != nil {
 		return err
 	}
 
-	for buf.Next() {
-		switch buf.MsgType {
+	var (
+		err  error
+		once sync.Once
+	)
+
+	doneReadingSignal := make(chan struct{}, 1)
+	done := func() { doneReadingSignal <- struct{}{} }
+
+	for r.Next(doneReadingSignal) {
+		switch r.MsgType {
 		case message.ServerHandshake:
 			// The client _MUST_ close the connection
 			// if the protocol version can't be supported.
 			// https://edgedb.com/docs/internals/protocol/overview
-			major := buf.PopUint16()
-			minor := buf.PopUint16()
+			major := r.PopUint16()
+			minor := r.PopUint16()
 
 			if major != 0 || minor != 8 {
-				if err := c.conn.Close(); err != nil {
-					return err
-				}
-
-				return fmt.Errorf(
-					"unsupported protocol version: %v.%v",
-					major,
-					minor,
+				e := fmt.Errorf(
+					"unsupported protocol version: %v.%v", major, minor,
 				)
+				return wrapAll(err, e, c.conn.Close())
 			}
 		case message.ServerKeyData:
-			buf.Discard(32) // key data
+			r.Discard(32) // key data
```

Comment:
I recommend implementing `discard_message`-like API. https://github.com/MagicStack/py-pgproto/blob/90ad4154a0aa6923e2f97d3d3d779c4b830ff415/buffer.pyx#L639-L655

BTW, what happens if I don't consume message data to its end? Or overread it?

Reply:
> BTW, what happens if I don't consume message data to its end? Or overread it?

The reader will panic. The r.Buf []byte is alwaysed caped so that over reading panics. If `len(r.Buf) > 0` then r.Next() panics.

Reply:
> I recommend implementing discard_message-like API

What benefit does that abstraction provide in this implementation? None of the checks/actions that the python implementation performs are required here.

Maybe this would be more maintainable if the length of the message changed in the future?
```go
r.Buf = r.Buf[len(r.Buf):]
```

Reply:
on second thought this probably be better
```go
r.Buf = nil
```

Reply:
> What benefit does that abstraction provide in this implementation?

You wouldn't need to count bytes by hand.

> r.Buf = nil

This is OK, but `r.DiscardMessage()` is way clearer to the reader of your code. It immediately tells the reader that you want to ignore the rest of the message, while `r.Buf = nil` requires the reader to understand your code way deeper.

The method would also allow you to tweak the actual implementation later, for which `r.Buf = nil` might be suboptimal.

In other words, compilers are smart enough to inline short helper functions like the proposed `DiscardMessage()`; we should use that fact when we can to write readable code.

Reply:
done :heavy_check_mark: